### ISSUE TITLE
handle-callback-err: does not differentiate method from variable

### ISSUE THREAD

**Issue by fxlemire** [buzinas/tslint-eslint-rules#153]
In the following error handling:

`error => console.error('Could not print the document');`

The error is not thrown because of console.**error**. But it should since the actual error variable is not handled.

relevant tslint config: {"handle-callback-err": [true, "^(err|error)$"]}

tslint: 4.0.2
tslint-eslint-rules: 3.2.0
typescript: 2.1.4
vscode: 1.7.2
vscode-tslint: 0.7.1

**Comment by jmlopez-rod** [comment]
Ah... I see, I'll do a fix this afternoon. Thanks.

**Comment by jmlopez-rod** [comment]
@fxlemire Is this valid?

```ts
var test = err => err.message;
```

The [eslint tests](https://github.com/eslint/eslint/blob/master/tests/lib/rules/handle-callback-err.js) say it is, but I want to say that it isn't since `err` could be `null`. The modification I made will flag your error but it would also mark the one I just mentioned. Here are a few tests:

```ts
  {
    code: `error => console.error('Could not print the document');`,
    options: ['error'],
    errors: [error]
  },
  {
    code: `error => something.something.error.something.err('Could not print the document');`,
    options: ['^(err|error)$'],
    errors: [error]
  },
  {
    code: `error => console.error(error.stack);`,
    options: ['^(err|error)$'],
    errors: [error]
  },
  {
    code: `error => error ? console.error(error.stack) : console.log('no error');`,
    options: ['^(err|error)$']
  }
```

Notice the last two examples, in the one we use `error.stack` by itself it flags it since `error` was not used by itself. Thoughts? Should we change

```ts
var test = err => err.message;
```

to be flagged?

**Comment by fxlemire** [comment]
@jmlopez-rod Interesting point.

My opinion would be that accessing a non-existent property on the error object is not the concern of this specific rule. As long as a user handles the error or one of its properties, it is fine.

I would agree with the eslint tests.

Another approach would be to keep your new interpretation by adding a flag that a user can add in his setting to make it stricter, e.g. `{"handle-callback-err": [true, "^(err|error)$", 'object-only']}`, which would force the user to handle the error object itself.

What do you think? Do you know anyone who should join this conversation?

**Comment by jmlopez-rod** [comment]
Funny that you mentioned it, I was thinking about the same idea this morning. I think I'll add an object option with the name you suggested, that way it is easier to turn it off and on in the configuration instead of having to remove the string.

```
{
  "handle-callback-err": [true, "^(err|error)$", {
    "objectOnly": true
  }]
}
```

@blakeembrey any thoughts on this?

**Comment by fxlemire** [comment]
The object is definitely the right approach. I'd make it optional, with `"objectOnly": false` being the default value.

False as the default, because given an error with the following properties: `status, message, time`, if a user handles `error.status` and `error.message`, the warning shouldn't pop up unless he explicitly stated that he wants this to be strict.

Adding on top of this example, if the user also handles `error.time`, would your implementation know that he handled all 3 properties? I presume not. So that's why I would go with false as default behavior.

Finally, I'm not necessarily fan of my own suggestion, didn't take time to think it through, haha! How about `allow-properties: true`? That would be my favorite.

Some of my other ideas were:

* `use-object-ref: false`
* `use-err-var: false`
* `no-properties: false`
* `use-properties: true`

**Comment by jmlopez-rod** [comment]
I agree, the default value would be false so that all the current tests may pass. It is also the current behavior with eslint.

I'm not completely sure about the option name. Let's give it a bit more thought. If this option is implemented we should give a more descriptive error as well:

Default error message:

```
Expected error to be handled.
```

When on strict usage mode:

```
Expected error to be handled without accessing its properties at least once.
```

**Comment by jmlopez-rod** [comment]
@fxlemire I opened a PR with the changes discussed here. Can you look over the tests I added and verify that they are correct? If there is no changes to be made I'll be making a release tomorrow night.

**Comment by fxlemire** [comment]
@jmlopez-rod added comments on the tests

**Comment by jmlopez-rod** [comment]
Thanks, both test you mentioned are meant to pass. The test that are expected to have errors define an array of errors. Any comments on the strict error message?

```
 'Expected error to be handled without property access at least once'
```

I also need to document this one change.

**Comment by fxlemire** [comment]
I think the strict error message is great, very succinct and descriptive! Added one more comment, please have a look.

**Comment by jmlopez-rod** [comment]
Could you take a look at the doc file: https://github.com/buzinas/tslint-eslint-rules/pull/157/files: src/docs/rules/handleCallbackErrRule.md.

**Comment by blakeembrey** [comment]
@jmlopez-rod Can you revert the release and invert this rule to make it stricter instead of starting strict. The release you did was a breaking change and broke projects like https://github.com/TypeStrong/ts-node/blob/42e5c1996c129688598755aa1286aecbd89872ef/src/_bin.ts#L413.

**Comment by jmlopez-rod** [comment]
Weird, that should have behave in the same way as the eslint rule. Let me check, just to be clear, we want

```
function isRecoverable (error: TSError) {
  return error.diagnostics.every(x => RECOVERY_CODES.indexOf(x.code) > -1)
}
```

to be ok, with `error` as the only parameter to the rule right?

**Comment by blakeembrey** [comment]
Yes, it should be passing. I might have misread the situation then, and it's a bug with the latest release in that case. I have a build failing with two instances, see https://travis-ci.org/TypeStrong/ts-node/builds/186493006.

Reference to ESLint because I just took a look: https://github.com/eslint/eslint/blob/f20b9e91b9ddc65418939ac8e242e2ff0dad5f02/lib/rules/handle-callback-err.js#L76

**Comment by jmlopez-rod** [comment]
Yes, just verified that it breaks with your example, let me come up with a fix.

**Comment by jmlopez-rod** [comment]
I see why it's breaking, but, it's dinner time so it'll be a while before I can release a fix, sorry for the inconvenience.
